plotter/contour.go

```diff
@@ -0,0 +1,488 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:generate ./list
+
+package plotter
+
+import (
+	"image/color"
+	"math"
+	"sort"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/palette"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// Contour implements the Plotter interface, drawing
+// a contour plot of the values in the GridXYZ field. The
+// row index is from top to bottom.
+type Contour struct {
+	GridXYZ GridXYZ
+
+	// Levels describes the contour heights to plot.
+	Levels []float64
+
+	// LineStyle is the style of the contour lines.
+	draw.LineStyle
+
+	// Palette is the color palette used to render
+	// the heat map.
+	Palette palette.Palette
+
+	// Underflow and Overflow are colors used to draw
+	// contours outside the dynamic range defined
+	// by Min and Max.
+	Underflow color.Color
+	Overflow  color.Color
+
+	// Min and Max define the dynamic range of the
+	// heat map.
+	Min, Max float64
+}
+
+// NewContour creates as new contour plotter for the given data,
+// using the provided palette.
```

Comment:
I don’t think hidden interface testing is ever obvious

It doesn’t need an extra type.

> On Feb 16, 2015, at 4:20 PM, Dan Kortschak <notifications@github.com> wrote:
> 
> In plotter/contour.go <https://github.com/gonum/plot/pull/15#discussion_r24784565>:
> 
> > +	// the heat map.
> > +	Palette palette.Palette
> > +
> > +	// Underflow and Overflow are colors used to draw
> > +	// contours outside the dynamic range defined
> > +	// by Min and Max.
> > +	Underflow color.Color
> > +	Overflow  color.Color
> > +
> > +	// Min and Max define the dynamic range of the
> > +	// heat map.
> > +	Min, Max float64
> > +}
> > +
> > +// NewContour creates as new contour plotter for the given data,
> > +// using the provided palette.
> Mmm. I guess so. I would have thought it was pretty obvious, but I'll add a comment. I don't think it warrants an additional type in the godoc.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/gonum/plot/pull/15/files#r24784565>.
>